pandas/core/array_algos/take.py

```diff
@@ -130,7 +135,8 @@ def take_1d(
     """
     Specialized version for 1D arrays. Differences compared to take_nd:
 
-    - Assumes input (arr, indexer) has already been converted to numpy array / EA
+    - Assumes input array has already been converted to numpy array / EA
```

Comment:
can you put asserts for these conditions

Reply:
Not really, as it's the point of this function to *not* do those asserts, but do them a level higher up in the calling code (and to be clear, this would just give a TypeError or AttributeError, it's not that it would segfault or so in case wrongly typed arguments are passed).

Reply:
sure you can we have asserts all of the place to enforce internal conditions, that's the point

Reply:
So the place where this function is called already has a `indexer = ensure_int64(indexer, copy=False)`. So adding a `assert indexer.dtype == "int64"` on the next line is then superfluous.

Reply:
pls do the asserts. documenting this is good, but an assert is way better.

Reply:
Have you read my response? 
The point of the changes in this PR is that we don't need to do those assertions 1000 times on the same data, but only once in the code calling this function. 

And the code calling this function asserts that the indexer is int64. And ensures that the input is an ndarray or EA.

Reply:
does this actually change the perf? asserts should not do that
if so that is a different issue

your comments don't change my mind here

Reply:
Of course adding those asserts gives a performance overhead. Maybe not much for a single call, but the point is that we call this function *repeatedly* on partly the same input, and then this overhead adds up (eg in the unstack benchmark that I reference in the top post, the `take_1d` is called 100_000 times for a single unstack operation, while we can do those asserts a single time inside `unstack`). 
It's not necessarily the `assert` itself, but the actual check after the assert takes some time. For example, see the following call of `take_1d` using the current branch of this PR. With this PR, a single call (using data similarly as would be used in the referenced unstack operation) takes a bit more than 500ns:

```python
In [1]: from pandas.core.array_algos.take import take_1d

In [2]: from pandas.api.extensions import ExtensionArray

In [3]: arr = np.random.randn(10000)

In [4]: indexer = np.arange(100, dtype="int64")

In [5]: %timeit take_1d(arr, indexer, allow_fill=False)
561 ns ± 5.32 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [6]: %timeit assert isinstance(arr, (np.ndarray, ExtensionArray))
92.7 ns ± 1.01 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [7]: %timeit assert isinstance(indexer, np.ndarray)
70.1 ns ± 3.73 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [8]: %timeit assert indexer.dtype == "int64"
287 ns ± 1.72 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Doing all the assertions as I now documented them (ensuring it are actual arrays, ensuring the dtype of the indexer), would together give an almost 2x slowdown in this case. Only checking for ndarray would still give an almost 20% slowdown.

Reply:
(as another example, we don't actually check that indices are within bounds in take_nd/take_1d, which would actually silently give garbage data if passed wrong indexer, while a wrong dtype for the indexer even gives a normal error message)

Reply:
ok. note that maybe we should do release builds to avoid this issue generally. but i guess this is fine.